tree/ntuple/v7/inc/ROOT/RPageSinkBuf.hxx

```diff
@@ -0,0 +1,84 @@
+/// \file ROOT/RPageSinkBuf.hxx
+/// \ingroup NTuple ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \author Max Orok <maxwellorok@gmail.com>
+/// \date 2021-03-17
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2021, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RPageSinkBuf
+#define ROOT7_RPageSinkBuf
+
+#include <ROOT/RPageStorage.hxx>
+
+#include <memory>
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RPageSinkBuf
+\ingroup NTuple
+\brief Abstract sink that coalesces cluster column page writes
+*/
+// clang-format on
+class RPageSinkBuf : public RPageSink {
+private:
+   /// A buffered column
+   class RColumnBuf {
+   private:
+      std::pair<RPageStorage::ColumnHandle_t, std::vector<RPage>> fBuf;
+   public:
+      void BufferPage(RPageStorage::ColumnHandle_t columnHandle, const RPage &page) {
+         if (!fBuf.first) {
+            fBuf.first = columnHandle;
+         }
+         fBuf.second.emplace_back(page);
```

Comment:
The problem might be here, or rather in `RPage`, which has a void pointer for the actual page data. So we don't copy the buffer but only the pointer.